### PR TITLE
Annotate highlighted events with notification reason

### PR DESCRIPTION
**Pull request by kerryarchibald**
Fixes https://github.com/vector-im/element-web/issues/24927
With https://github.com/matrix-org/matrix-js-sdk/pull/3284
image

For highlighted messages, adds a notification icon to the message action bar explaining the highlight in a tooltip.
(May be expanded on later to [add a dialog with rule management/anti-abuse options](https://github.com/vector-im/element-web/issues/24927#issuecomment-1498506611))

- Adds test util for default push rules and push rule creation

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Annotate highlighted events with notification reason ([\#10599](https://github.com/matrix-org/matrix-react-sdk/pull/10599)). Fixes vector-im/element-web#24927. Contributed by @kerryarchibald.

### CODE REVIEW

File: src/utils/event/highlight.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    if (rule.rule_id === RuleId.AtRoomNotification || rule.rule_id === RuleId.IsRoomMention) {
+        return _t("This message mentions the room.");
+    }
+    if (rule.rule_id === RuleId.IsUserMention) {
+        return _t("You were explicitly mentioned.");
+    }
+    if (rule.kind === PushRuleKind.ContentSpecific) {
+        return _t(`Your keyword '%(keyword)s' was mentioned.`, { keyword: rule.pattern! });
+    }
+    return "";
```
Comment (justjanne) [inline src/utils/event/highlight.ts:107]: This should probably be a switch-case, at least [according to the code style](https://github.com/vector-im/element-web/blob/develop/code_style.md?plain=1#L127)

File: src/components/views/messages/MessageActionBar.tsx [hunk 2]
```diff
@@ -579,6 +582,25 @@ export default class MessageActionBar extends React.PureComponent<IMessageAction
...
+            const reason = getHighlightReasonMessage(highlightInfo.rule);
+            toolbarOpts.splice(
+                0,
+                0,
+                <TooltipTarget key="highlight-annotation" data-testid="highlight-annotation" label={reason}>
+                    <div className="mx_MessageActionBar_infoElement">
+                        <NotificationIcon />
+                    </div>
+                </TooltipTarget>,
+            );
```
Comment (justjanne) [inline src/components/views/messages/MessageActionBar.tsx:601]: While not necessarily intuitive, you might want to use `Array.unshift` here.

| -         | Add     | Remove |
| --------- | ------- | ------ |
| **Start** | unshift | shift  |
| **End**   | push    | pop    |

**Review by justjanne — APPROVED** [review]
Really good code! I've only got some optional comments regarding code style, but those are not necessarily any "better", just different.

**Comment by daniellekirkwood** [comment]
On first impression I really like this from a user perspective but the more I played with it the more I realised that it's quite odd to have a hover icon in the action bar... All other actions there have a click thru but the bell is a hover over (which might not be a11y friendly?)

As folks often miss the "(edited)" text is a button I wonder if we can combine the two things here... So we could pick a slightly different icon, on click the Edited modal appears. The hover over of the button would still tell users why they were pinged, then clicking on it will show the edits. Alternatively maybe hovering over the "(edited)" text shows the tool tip? 

Either way, I think we need some design input here: @rufuskahler could you help us out?
The problem we're trying to solve is: The user doesn't know why the message text is red. It could be red for any number of reasons. Eg: Your keyword was mentioned... Your display name was mentioned...

**Review by justjanne — APPROVED** [review]

**Comment by rufuskahler** [comment]
This is a really nice implementation @kerryarchibald – However, I have my doubts about using the action bar for the reasons Danielle mentioned. Furthermore, I'd like to discuss other ways of solving the problem closer to the root cause, e.g with Compound, we're moving away from red highlighted text for notifications. @daniellekirkwood – This sounds like more consensus is needed between Product & Design.

**Pull request closed without merging**